File: week-12/grep.pl
#!/usr/bin/perl

use strict;
use warnings;

sub grep_in_directory {
    my ($directory, $pattern) = @_;
    my $count = 0;

    eval {
        opendir(my $dh, $directory) or die "Cannot open directory $directory: $!";
        while (my $file = readdir($dh)) {
            next if $file =~ /^\./;
            my $file_path = "$directory/$file";
            if (-f $file_path) {
                open my $fh, '<', $file_path or die "Cannot open file $file_path: $!";
                while (my $line = <$fh>) {
                    if ($line =~ /$pattern/) {
                        $count++;
                    }
                }
                close $fh;
            } elsif (-d $file_path) {
                $count += grep_in_directory($file_path, $pattern);
            }
        }
        closedir($dh);
    };

    if ($@) {
        warn "Error processing directory $directory: $@";
    }

    return $count;
}

sub main {
    my @directories;
    my $current_pattern;

    my $i = 0;
    while ($i < @ARGV) {
        if ($ARGV[$i] eq "-d") {
            $i++;
            if ($i < @ARGV) {
                my $directory = $ARGV[$i];
                push @directories, $directory;
                $current_pattern = undef;
            } else {
                die "Error: Missing directory after -d\n";
            }
        } else {
            $current_pattern = $ARGV[$i];
            foreach my $directory (@directories) {
                my $count = grep_in_directory($directory, $current_pattern);
                print "$directory: $count matches for '$current_pattern'\n";
            }
        }
        $i++;
    }
}

main();

File: week-12/grep.py
#!/usr/bin/env python3

import os
import sys

def grep_in_directory(directory, pattern):
    count = 0
    try:
        for root, dirs, files in os.walk(directory):
            for file in files:
                file_path = os.path.join(root, file)
                with open(file_path, 'r', errors='ignore') as f:
                    for line in f:
                        if pattern in line:
                            count += 1
    except Exception as e:
        print(f"Error processing directory {directory}: {e}")
    
    return count

def main():
    directories = []
    current_pattern = None

    i = 1
    while i < len(sys.argv):
        if sys.argv[i] == "-d":
            i += 1
            if i < len(sys.argv):
                directory = sys.argv[i]
                directories.append(directory)
                current_pattern = None
            else:
                print("Error: Missing directory after -d")
                sys.exit(1)
        else:
            current_pattern = sys.argv[i]
            for directory in directories:
                count = grep_in_directory(directory, current_pattern)
                print(f"{directory}: {count} matches for '{current_pattern}'")
        i += 1

if __name__ == "__main__":
    main()
